src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/IONFILEUriHelper.kt

```diff
@@ -0,0 +1,173 @@
+
+    private val internalCacheDir = context.cacheDir
+    private val internalFilesDir = context.filesDir
+    private val externalCacheDir = context.externalCacheDir
+    private val externalFilesDir = context.getExternalFilesDir(null)
+
+    /**
+     * Resolves a URI for a file.
+     *
+     * Identifies the URI as belonging to a local file, or a file with content:// scheme
+     *
+     * @param unresolvedUri the URI to resolve; see [IONFILEUri.Unresolved]
+     * @return success with the resolved URI [IONFILEUri.Resolved], or error otherwise
+     */
+    suspend fun resolveUri(
+        unresolvedUri: IONFILEUri.Unresolved
+    ): Result<IONFILEUri.Resolved> = runCatching {
+        val parentFolderObject = unresolvedUri.parentFolder.getFolderFileObject()
+        val resolvedUri = if (parentFolderObject == null) {
+            val parsedUri = Uri.parse(unresolvedUri.uriPath)
+            when {
+                parsedUri.scheme == CONTENT_SCHEME_NAME -> resolveAsContentUri(parsedUri)
+
+                unresolvedUri.uriPath.contains(SYNTHETIC_URI_PREFIX) ->
+                    resolveAsContentUri(convertSyntheticToContentUri(unresolvedUri.uriPath))
+
+                parsedUri.scheme == FILE_SCHEME_NAME -> resolveAsLocalFile(
+                    parentFolderFileObject = null,
+                    parsedUri.path ?: ""
+                )
+
+                parsedUri.scheme == null -> resolveAsLocalFile(
+                    parentFolderFileObject = null,
+                    unresolvedUri.uriPath
+                )
+
+                else -> throw IONFILEExceptions.UnresolvableUri(unresolvedUri.uriPath)
+            }
+        } else {
+            resolveAsLocalFile(
+                parentFolderObject,
+                unresolvedUri.uriPath,
+                assumeExternalStorage = unresolvedUri.parentFolder?.inExternalStorage
+            )
+        }
+        resolvedUri
+    }
+
+    /**
+     * Resolves as a content:// URI
+     *
+     * @param uri the content:// URI
+     * @return a [IONFILEUri.Resolved.Content] object
+     */
+    private fun resolveAsContentUri(uri: Uri): IONFILEUri.Resolved.Content {
+        return IONFILEUri.Resolved.Content(uri)
+    }
+
+    /**
+     * Resolves to a local file URI
+     *
+     * @param parentFolderFileObject the parent folder of the file, or null if there is none
+     * @param localPath the local path to the file (minus the parent folder path)
+     * @param assumeExternalStorage true if file is in external storage, false if it does not,
+     *  null if unknown (will be determined from file path)
+     * @return a [IONFILEUri.Resolved.Local] object
+     */
+    private suspend fun resolveAsLocalFile(
+        parentFolderFileObject: File?,
+        localPath: String,
+        assumeExternalStorage: Boolean? = null
+    ): IONFILEUri.Resolved.Local = withContext(Dispatchers.IO) {
+        val localFileObject = if (parentFolderFileObject != null) {
```

Comment:
Not super important but we could improve readability by extracting this to a separate function:

`private fun getLocalFileObject(parentFolderFileObject: File?, localPath: String): File {
    return parentFolderFileObject?.let { File(it, localPath) } ?: File(localPath)
}`